### ISSUE TITLE
Break out each Read physical operation into its own RPC call

### ISSUE THREAD

**Issue by nathanielc** [influxdata/platform#76]
Currently we have a Read call with various parameters. Not all combinations of parametes are valid, and certain combinations require that the read service behave differently. 

By explicitly exposing those individual physical operations we prevent invalid combinations, expose to the planner the cost of each operation, protect against breaking existing behavior and separate concerns between the client and the server.

**Comment by wolffcm** [comment]
It might be good to think of operations as being applied in a particular order by storage, and letting the planner attempt to put all the operations in that order, so that they can be pushed down into storage in a way that avoids a combinatorial explosion of logic.

Consider that storage will perform the following operations in order:
1. range
2. filter by measurement, field or tag key
3. group (by anything except `_value` or `_time`)
4. aggregate

We can think of them as a stack of operations, with `range` being on the bottom and `aggregate` being on the top.

Given this there might be 4 different RPC calls:
1. ReadRange
2. ReadFilter (performs range, filter)
3. ReadGroup (performs range, filter and group)
4. ReadAggregate (performs range, filter, group and aggregate)

There would then be four possible kinds of `FromProcedureSpec` that correspond to the RPC calls.  There should be a 1:1 relationship there.

The job of the logical planner would be to put operations into this order if possible.  So a Flux query that was `from |> group |> filter |> range |> mean` would be rewritten into `from |> range |> filter |> group |> mean`.

The physical planner would then push down operations in the order provided by the logical planner, and the operation that's currently "on top" in the InfluxDB source would determine what can be pushed down.  So for example, if we have already pushed down an aggregation like `mean`, we couldn't push down a `filter` on top of that---it would need to be evaluated in Flux.  If the filter could be pushed down below `mean`, the logical planner would have already done that.

This is just the basic idea.  There are other details to iron out:
- Meta queries should get their own RPC call
- What to do about limiting operations like `PointsLimit`, `SeriesLimit`?
- Windowing?
- etc

 @stuartcarnie This is my summary of our conversation from earlier today.  Feel free to add anything that I've missed.

**Comment by stuartcarnie** [comment]
Great summary, @wolffcm. This aligns with my understanding of our conversation 👍

**Comment by stuartcarnie** [comment]
Presumably for a query `from |> range |> group`, `ReadGroup` would be used, but the filter predicate is optional?

**Comment by wolffcm** [comment]
Yes, that's what I'm thinking.  For a "stack" of operations, the one on the bottom (always `range`) will be required, as will the one on the top.  The ones on the middle may be optional or represented trivially (e.g., a filterless `readGroup` could be thought of as having a simple `true` predicate).

**Comment by nathanielc** [comment]
This is great! It provides a clear model to work with.


As for some of the questions. 

> Meta queries should get their own RPC call

Agreed, do meta queries have their own "stack"?

> What to do about limiting operations like PointsLimit, SeriesLimit?

My understanding is that these get applied at some stage in the process, so maybe we just need a Limit step in the stack? Not sure where that lives.

> Windowing?

To keep storage simple it will never manage windowing itself. Meaning all storage supports is range queries. If the planner wants to issue one large read to get lots of data and then window it on the query engine side it can. Alternantively the planner can submit an individual read range per window. The planner gets to decide and storage has no concept of windows.

**Comment by jlapacik** [comment]
Does each RPC call represent a unique access pattern? For instance, it seems like the disk access pattern for `ReadFilter` would be different depending on whether the filter references a field (file scan) or a tag (index scan). Should these be separate RPC calls/physical procedures?

**Comment by stuartcarnie** [comment]
@jlapacik I think that would become unwieldy, as it would require three permutations of all APIs to account for:

* filter only tags
* filter only values
* filter tags and values

However, I agree that the planner should be able to query this cost to determine the most optimal plan. Perhaps in the future, we pair the physical operations with APIs to estimate the cost of the physical operations?

For example:

* `ReadFilterEstimateCost`
* `ReadGroupEstimateCost`

where these APIs return estimates for the number of reads, etc

**Comment by wolffcm** [comment]
@stuartcarnie  We could separate out filtering tags from filtering values in the operation stack so it could become:
1. range
2. filter tags
3. filter values
4. group
5. aggregate

The planner would be able to make this distinction based on the column names.  I'm fine with either way, just throwing it out there.

I like the idea of RPC calls that we can use to get stats from...

**Comment by wolffcm** [comment]
Yes, that would complicate things a bit.  We could redefine our terms to have two kinds of filters:
1. Filters that select series by only looking at tag keys
2. Filters that must look at the values in the series (potentially in addition to tag keys)

It tends to most useful to break a filter up into components that are joined by `AND`.  In the example you gave, the top-level operator is `OR`, so it would be treated atomically and fall into the second category.

If the planner can do some analysis that simplifies how storage fulfill requests, it might be worth exploring.
